lib/hexpm/accounts/web_auth.ex

```diff
@@ -0,0 +1,117 @@
+defmodule Hexpm.Accounts.WebAuth do
+  use Hexpm.Context
+
+  @moduledoc false
+
+  # A pool for storing and validating web auth requests.
+
+  alias Hexpm.Accounts.WebAuthRequest, as: Req
+  alias Hexpm.Accounts.Keys
+
+  # `device_code` refers to the code assigned to a client to identify it
+  # `user_code` refers to the code the user enters to authorize a client
+  # `scope` refers to the permissions granted to a key
+  # `scopes` refers to the list of scopes that are allowed in a web auth request
+  # `key_permission` is the default permissions given to generate a hex api key
+  # `key_params` is the default params to generate a hex api key
+
+  @key_permission %{domain: "api", resource: nil}
+  @key_params %{name: nil, permissions: [@key_permission]}
+
+  @doc """
+  Adds a web auth request to the pool and returns the response.
+
+  ## Params
+
+    - `scope` - Scope of the key to be generated. One of read and write.
+  """
+
+  def get_code(scope) do
+    device_code = "foo"
+    user_code = "bar"
+
+    request = %{
+      device_code: device_code,
+      user_code: user_code,
+      scope: scope,
+      verified: false
+    }
+
+    case Req.changeset(%Req{}, request) |> Repo.insert() do
+      {:ok, _req} ->
+        %{
+          device_code: device_code,
+          user_code: user_code
+        }
+
+      error ->
+        error
+    end
+  end
+
+  @doc """
+  Submits a verification request to the pool and returns the response.
+
+  ## Params
+
+    - `user` - The user for whom the key should be generated.
+    - `user_code` - The user code entered by that user.
+    - `audit` - Audit data for generating the key.
+  """
+  def submit(user, user_code, audit) do
+    request = Req |> Repo.get_by(user_code: user_code)
+
+    if request do
+      {_user, audit_con} = audit
+
+      change = %{verified: true, user_id: user.id, audit: audit_con}
+
+      Req.changeset(request, change) |> Repo.update()
```

Comment:
if the DB call fails, our function returns `{:error,%Changeset{} | "invalid user code"}`. Can we do something more uniform, ie we always return a changeset or always a string error?

Reply:
@wojtekmach, which do you prefer? Changeset or string?

Reply:
whatever is simpler to implement, probably a string.